storage-node/src/cache/data_store_cache/sqlite/mod.rs

```diff
@@ -0,0 +1,252 @@
+{
+    replacer: RwLock<R>,
+    sqlite_base_path: String,
+    buffer_size: usize,
+}
+
+impl<R: DataStoreReplacer<SqliteStoreReplacerKey, SqliteStoreReplacerValue>> SqliteStoreCache<R> {
+    pub fn new(
+        replacer: R,
+        sqlite_base_path: String,
+        buffer_size: Option<usize>,
+    ) -> ParpulseResult<Self> {
+        let db = Connection::open(&sqlite_base_path)?;
+        let create_table_stmt = format!(
+            "CREATE TABLE IF NOT EXISTS {} ({} BLOB);",
+            SQLITE_CACHE_TABLE_NAME, SQLITE_CACHE_COLUMN_NAME
+        );
+        let buffer_size = buffer_size.unwrap_or(SQLITE_BLOB_READER_DEFAULT_BUFFER_SIZE);
+        db.execute_batch(&create_table_stmt)?;
+
+        Ok(Self {
+            replacer: RwLock::new(replacer),
+            sqlite_base_path,
+            buffer_size,
+        })
+    }
+}
+
+impl<R: DataStoreReplacer<SqliteStoreReplacerKey, SqliteStoreReplacerValue>> Drop
+    for SqliteStoreCache<R>
+{
+    fn drop(&mut self) {
+        fs::remove_file(self.sqlite_base_path.clone()).expect("remove sqlite db files failed");
+    }
+}
+
+#[async_trait]
+impl<R: DataStoreReplacer<SqliteStoreReplacerKey, SqliteStoreReplacerValue>> DataStoreCache
+    for SqliteStoreCache<R>
+{
+    async fn get_data_from_cache(
+        &mut self,
+        remote_location: String,
+    ) -> ParpulseResult<Option<Receiver<ParpulseResult<Bytes>>>> {
+        let mut replacer = self.replacer.write().await;
+        if let Some(replacer_value) = replacer.get(&remote_location) {
+            let (tx, rx) = channel(SQLITE_BLOB_CHANNEL_CAPACITY);
+            let row_id = *replacer_value.as_value();
+            let sqlite_base_path = self.sqlite_base_path.clone();
+            let buffer_size = self.buffer_size;
+
+            tokio::spawn(async move {
+                let db =
+                    Connection::open_with_flags(sqlite_base_path, OpenFlags::SQLITE_OPEN_READ_ONLY)
+                        .unwrap();
+                let mut blob_reader = SqliteBlobReader::new(&db, row_id, buffer_size).unwrap();
+                while let Some(result) = blob_reader.next() {
+                    match result {
+                        Ok(bytes_read) => {
+                            let buffer = blob_reader.buffer();
+                            let bytes = Bytes::copy_from_slice(&buffer[..bytes_read]);
+                            tx.send(Ok(bytes)).await.unwrap()
+                        }
+                        Err(err) => tx.send(Err(err)).await.unwrap(),
+                    }
+                }
+            });
+            Ok(Some(rx))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn put_data_to_cache(
+        &mut self,
+        remote_location: String,
+        data_size: Option<usize>,
+        mut data_stream: StorageReaderStream,
+    ) -> ParpulseResult<usize> {
+        let mut replacer = self.replacer.write().await;
```

Comment:
So actually replacer acts as a large lock to basically lock all... But is it okay maybe, otherwise things may get complicated...

Reply:
Let's first write a simple version and optimize it later 🥹

Reply:
Sure!!